sudoku example notebook

mjhajharia opened the pull request:
this PR adds a notebook that explains and visualizes sudoku as a graph coloring problem using networkx

Reference: https://github.com/networkx/networkx/discussions/5085#discussioncomment-1435440

from #52 

cc: @dschult @rossbar

rossbar commented:
Thanks for being flexible and adjusting the PR to better fit the workflow @mjhajharia ! I've gone ahead and added your new notebook to the generators `toctree`, so the CI should pass now and new page should show up [in the `build-docs artifact`](https://108-40210422-gh.circle-artifacts.com/0/site/_build/html/content/generators/sudoku.html).

I've only had a chance to look briefly over it, but it looks really nice! I'll try to do a more detailed review asap. Thanks again!

rossbar commented:
Thanks for this notebook @mjhajharia , the visualizations in particular are really great!

I went ahead and made another change just to get things better integrated into the workflow here: rather than pre-compute and store the images as png's, I've updated the notebook so that it is executed directly as part of the build process, so the images are generated by the code itself each time. One of the advantages of this approach (aside from the fact that it saves storage space) is that it helps us guarantee that the code works as expected as things change in the future.

I made these changes in 9a3126d - I didn't touch the content itself, so everything should look the same in the final website: https://130-40210422-gh.circle-artifacts.com/0/site/_build/html/content/generators/sudoku.html

rossbar commented on content/generators/sudoku.md:
````diff
@@ -0,0 +1,163 @@
+sudoku(3, 'grid')
+```
+
+We couldn't see a lot of edges in the first figure because they fell on top of each other, this is perhaps a better view, if you want to look at this in an animated way, check [this old reddit post](https://reddit.com/r/dataisbeautiful?utm_content=subreddit&utm_medium=post_embed&utm_name=8775be457c4149bd9d15290fa5723658&utm_source=embedly&utm_term=6ty4vf) although the code isn't available there a ...
````
I'm -1 on linking to the old reddit post. While the visualization is nice, it doesn't have a lot of pedagogical value. My vote would be to remove it for now, or better yet - try a tool like [bokeh](https://bokeh.org/) to implement the interactive visualization. If you were interested in doing so, there's [an example in the bokeh gallery that might prove useful](https://docs.bokeh.org/en/latest/docs/gallery/network_graph.html)

rossbar commented on content/generators/sudoku.md:
```diff
@@ -0,0 +1,163 @@
+    return box_edges, row_edges, column_edges
+
+# We have the edges and the nodes, now we want a networkx graph
+def draw_networkx_graph(n, board, box_edges, row_edges, column_edges, layout='grid'):
```
I think it would be an improvement to alter the name of this function slightly, only because it is *very* similar to some of the drawing function names in `nx_pylab`. Maybe `draw_sudoku` or `visualize_sudoku_graph`?

Ultimately I think it'd also be an improvement to move some of the processing done in the function (i.e. the graph building and layout handling) outside of the function, but that will be easier to discuss in a follow-up PR IMO, so I wouldn't worry about it now!

rossbar reviewed (commented):
This generally looks good, thanks @mjhajharia . I had a couple suggestions/comments after reading through everything, but nothing major.

One thing that stands out is that the notebook doesn't really have much connection to the networkx function for generating sudoku graphs `nx.sudoku_graph`. It would be nice to add something to tie back the notebook example to the generator in NetworkX. In fact, I think there's a case for some of the functionality here to be added to `networkx.generators.sudoku` if you were interested in making a contribution to NetworkX itself!

mjhajharia commented:
Hi! @rossbar @MridulS I'm so sorry this PR completely slipped my mind and was buried somewhere in Github notifications, I checked it today and made changes. So i made a lot of changes actually. Since ross mentioned trying to use the actual networkx graph for visualization instead of a custom graph that I generate, which I think is a good idea. So I spent some time thinking today about inferring "box edges", "row edges" and "column edges" from a graph, it turned out to be pretty neat. I can't exactly recall but I really think there's proper theoretical backing behind why it came out neatly (it mostly happened because Graph nodes were 0-80 in the default graph) I've added a lot of comments to clear things up but I think more polishing will need to be done so just tag me or something when that's needed. I had a very fun time doing this today, additionally I made a PR in networkx Repo as well as ross suggested, I don't know if it's good enough for the code repo so feel free to go with this or that one and I'll close one of them accordingly.

mjhajharia commented:
Reference to understand the code with comments -> https://nbviewer.org/gist/mjhajharia/5a9ede19745f54f8b390680bff1a0439 it's hacky but yeah I've explained almost everything here

mjhajharia commented:
btw a suggestion: not sure what's used for documentation of nx-guides currently, but here at pymc we're doing these couple of events and made executable notebooks, so we decided to to use thebe which works nicely with jupyter-notebooks and renders them as executables, there's no need to even upload a .md file we simply upload notebooks https://pymc-data-umbrella.xyz/en/latest/webinars/probabilistic_programming_with_pymc/notebook.html#probprog-pymc-nb which is easier!!

mjhajharia commented:
so i did a couple of stupid things and then realised them, now I changed them and updated the notebook to the best of my abilities!! I'm sorry for the delay, although I'm afraid I might not be active here to do finishing touches if needed because of other work but here's a link for the notebook I added, I think it works now! https://nbviewer.org/gist/mjhajharia/bed5e13ab84660456323648d6d99d2fd

MridulS commented:
Thanks @mjhajharia! I made a couple of changes and did some code formatting.

Maybe we should also add a section on using `nx.greedy_color` to solve a sudoku puzzle too! (But we can't start with a partially colored graph :/)

mjhajharia commented:
> Thanks @mjhajharia! I made a couple of changes and did some code formatting.
> 
> Maybe we should also add a section on using `nx.greedy_color` to solve a sudoku puzzle too! (But we can't start with a partially colored graph :/)

yes i did try that actually!! it works fine with saturation_largest_first, but yeah same thing in this example i started with a semi-filled one so its not that useful, but it can be used to find a random solved sudoku, like on the base structure of the graph of a sudoku if you want i can add that

rossbar reviewed (approved):
LGTM, thanks @mjhajharia for the very nice tutorial!

> btw a suggestion: not sure what's used for documentation of nx-guides currently, but here at pymc we're doing these couple of events and made executable notebooks, so we decided to to use thebe which works nicely with jupyter-notebooks and renders them as executables, there's no need to even upload a .md file we simply upload notebooks https://pymc-data-umbrella.xyz/en/latest/webinars/probabilistic_programming_with_pymc/notebook.html#probprog-pymc-nb which is easier!!

**Warning - meta discussion below, completely unrelated to the PR :)**

Thanks for the suggestion - for the design of the tutorials site, we decided to go with something that supports both executable notebooks (for working interactively) and fully-feature static sites which are typically really valuable for organizing and communicating information. Another important goal is to have a solution that works for collaborative authoring, i.e. we wanted something that would work with the traditional version control/GH workflow so that community authored content can be tested and reviewed. We decided to go with the tools from the [executablebooks project](https://github.com/executablebooks), which is also the organization that maintains thebe! [Thebe integration](https://myst-nb.readthedocs.io/en/latest/examples/interactive.html?highlight=thebe#widgets-and-interactive-outputs) is definitely a nice feature, though to this point we've stuck with binder integration since that was the tool the maintainers are the most familiar with. If you're interested in looking into incorporating thebe integration into the tutorials site I'd be interested to see your ideas!
